verify_rpms/rpm_verifier.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+
+"""Verify RPMs are signed"""
+
+import tempfile
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Callable, Iterable
+
+import click
+
+
+@dataclass(frozen=True)
+class ProcessedImage:
+    """Unsigned RPMs for a single image"""
+
+    image: str
+    unsigned_rpms: list[str]
+
+
+@dataclass
+class RPM:
+    """RPM name and whether it's signed"""
+
+    name: str
+    signed: bool
+
+
+def get_rpmdb(container_image: str, target_dir: Path) -> Path:
+    """Extract RPM DB from a given container image reference"""
+    raise NotImplementedError()
+
+
+def get_rpms(rpmdb: Path) -> list[RPM]:
+    """Get RPMs from RPM DB path"""
+    raise NotImplementedError()
+
+
+def generate_output(
+    unsigned_rpms: Iterable[ProcessedImage], fail_unsigned: bool
+) -> None:
+    """
+    Given a mapping between images and unsigned RPMs, print out the images
+    having unsigned RPMs, together with their unsigned RPMs.
+    Exit with error in case unsigned RPMs found and the relevant flag provided.
+    """
+    raise NotImplementedError()
+
+
+@dataclass(frozen=True)
+class ImageProcessor:
+    """Find unsigned RPMs provided an image"""
+
+    db_getter: Callable[[str, Path], Path] = get_rpmdb
+    rpms_getter: Callable[[Path], list[RPM]] = get_rpms
+
+    def __call__(self, img: str) -> ProcessedImage:
```

Comment:
I'm not convinced this unit of logic is significant enough to necessitate injecting dependencies into it and testing it in isolation.

Reply:
That is fine. But a coarse test that passes an actual image would be sufficient. Trying to mock-up the lower level elements is not beneficial in this case IMO. The core of the code is the "oc" and "rpm" calls, mocking them out makes into a pretty boring a useless glue code that does not IMO merit its own tests.

Reply:
How can we unit test without mocking the `oc` and `rpm` calls?

Reply:
Change our perception of what we consider to be the "unit" in this case. Note that is a "unit" test not "function" test or "python code" test. In this case the unit is the whole functional element, including the 3rd party code bits that happen to be invoked via process invocation.

Since we can arrange for both RPM and OC to perform no significant amount of IO or network traffic by E.g. inspecting a local image. I don't think mocking them out would significantly improve the test or the overall code.

Reply:
Are you suggesting we use the actual binaries? This means harder onboarding for both developers and CI.
I wouldn't want to have to install new binaries to the container we're using for CI just to be able to run new unit tests.

Reply:
"rpm" and "oc" aught to be commonplace enough to make us not worry too much about the developers. WRT the test container - if we base it on the actualy container we release it would have the tools there already anyway.

Not running the binaries makes are tests basically test nothing. In other words, we don't care that we run 'oc/rpm' with particular parameters, we care that we manage to make then do the particular tasks we need them to do.

Reply:
> "rpm" and "oc" aught to be commonplace enough to make us not worry too much about the developers. WRT the test container - if we base it on the actual container we release it would have the tools there already anyway.

I'd prefer not to have to get this one solved before we can merge this PR - or finish the entire epic/feature/outcome for that matter :) (our CI is using ubuntu at the moment, so it has neither of those packages).

> Not running the binaries makes are tests basically test nothing. In other words, we don't care that we run 'oc/rpm' with particular parameters, we care that we manage to make them do the particular tasks we need them to do.

That makes our unit tests depend on many different things which are not the code. It will take a lot of work for this not to be flaky (I appreciate that this might represent the fragility of the solution and not just the tests).

All in all, I don't see testing with the actual binaries something that I'd like to achieve before we can deliver this feature. I understand that the current `test_main` approach has its drawbacks. The alternative I have at the moment for the test implementation is to delete the test (I'd prefer keeping it, as I think it does serve some value).